### PR TITLE
[libc++] Simplify the implementation of LWG2762 (noexcept for unique_ptr)

### PR DESCRIPTION
**Pull request by ldionne**
I had originally made some comments in https://reviews.llvm.org/D128214 that were followed when we implemented LWG2762. I don't understand why I made these comments anymore, but either way it seems like I was wrong since using `unique_ptr<void>::operator*` should be ill-formed. All other implementations also make that ill-formed.

### CODE REVIEW

**Comment by mordante** [comment]
IIRC there were issue to test `static_assert(noexcept(unique_ptr<void>::operator*()));` But let's see what the CI says.

File: libcxx/test/std/utilities/memory/unique.ptr/noexcept_operator_star.compile.pass.cpp [hunk 1]
```diff
@@ -1,25 +0,0 @@
...
-// UNSUPPORTED: c++03
-
-// unique_ptr
-
-// add_lvalue_reference_t<T> operator*() const noexcept(noexcept(*declval<pointer>()));
-
-// Dereferencing pointer directly in noexcept fails for a void pointer.  This
-// is not SFINAE-ed away leading to a hard error. The issue was originally
-// triggered by
-// test/std/utilities/memory/unique.ptr/iterator_concept_conformance.compile.pass.cpp
```
Comment (mordante) [inline libcxx/test/std/utilities/memory/unique.ptr/noexcept_operator_star.compile.pass.cpp:18]: This was the cause for adding this.

**Comment by ldionne** [comment]
> IIRC there were issue to test `static_assert(noexcept(unique_ptr<void>::operator*()));` But let's see what the CI says.

My understanding is that `static_assert(noexcept(unique_ptr<void>::operator*()));` is ill-formed. All other implementations produce a hard-error for that case, and that seems to be what the Standard specifies.

File: libcxx/test/std/utilities/memory/unique.ptr/noexcept_operator_star.compile.pass.cpp (hunk 1 above)
Comment (ldionne) [reply libcxx/test/std/utilities/memory/unique.ptr/noexcept_operator_star.compile.pass.cpp:18]: Got it. FWIW this test passes after applying this patch locally. Perhaps it used to fail with an older compiler? I seem to recall older Clang instantiating `noexcept` too eagerly at some point in the past.

**Review by mordante — APPROVED** [review]
LGTM with the CI passing.

**Comment by frederick-vs-ja** [comment]
It seems that the SFINAE-friendness of `unique_ptr<void>::operator*` was originally expected and tested in https://reviews.llvm.org/D100073 (57ebf3d0087a8f988fac2409d04c953b850cb05f). Should we just remove the test coverage?

**Comment by ldionne** [comment]
> It seems that the SFINAE-friendness of `unique_ptr<void>::operator*` was originally expected and tested in https://reviews.llvm.org/D100073 ([57ebf3d](https://github.com/llvm/llvm-project/commit/57ebf3d0087a8f988fac2409d04c953b850cb05f)). Should we just remove the test coverage?

Yes, I think that makes the most sense.

**Pull request merged**